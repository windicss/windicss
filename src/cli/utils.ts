import fs from 'fs';
import path from 'path';
import glob from 'glob';
import minimatch from 'minimatch';

export function globArray(patterns: string[], options?: glob.IOptions): string[] {
  const list: string[] = [];

  patterns.forEach(pattern => {
    if (pattern[0] === '!') {
      let i = list.length-1;
      while(i > -1) {
        if (!minimatch(list[i], pattern)) {
          list.splice(i, 1);
        }
        i--;
      }

    }
    else {
      const newList = glob.sync(pattern, options);
      newList.forEach(item => {
        if (list.indexOf(item)===-1) {
          list.push(item);
        }
      });
    }
  });

  return list;
}

export function getVersion(): string {
  // eslint-disable-next-line quotes
  return `__NAME__ __VERSION__`; // replace by rollup
}

export function fuzzy(content: string): string[] {
  return content.match(/([^<>"'`\s]*[^<>"'`\s:])|([^<>"'`\s.(){}[\]#=%]*[^<>"'`\s.(){}[\]#=%:])/g) ?? [];
}

export function generateTemplate(
  folder: string,
  outputPath = 'windi.css'
): { html: string; css: string } {
  if (!(fs.existsSync(folder) && fs.lstatSync(folder).isDirectory())) {
    fs.mkdirSync(folder);
    if (!fs.existsSync(folder))
      throw new Error(`Folder ${folder} creation failed.`);
  }
  folder = path.resolve(folder);
  const template = `<!DOCTYPE html>
  <html lang="en">

  <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>${path.basename(folder)}</title>
      <link rel="stylesheet" href="${outputPath}">
  </head>

  <body class="bg-gray-100">
      <div class="container mx-auto flex flex-col justify-center items-center h-screen">
          <div class="lg:flex shadow rounded-lg">
              <div class="bg-blue-500 rounded-t-lg lg:rounded-tr-none lg:rounded-l-lg lg:w-4/12 py-4 h-full flex flex-col justify-center">
                  <div class="text-center tracking-wide">
                      <div class="text-white font-bold text-8xl lg:text-4xl">24</div>
                      <div class="text-white font-normal text-4xl pt-4 lg:pt-0 lg:text-2xl">Sept</div>
                  </div>
              </div>
              <div class="w-full px-1 bg-white py-5 lg:px-2 lg:py-2 tracking-wide">
                  <div class="flex flex-row lg:justify-start justify-center">
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          1:30 PM
                      </div>
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          Organiser : IHC
                      </div>
                  </div>
                  <div class="text-gray-700 text-xl pb-1 text-center lg:text-left px-2">
                      International Conference Dubai
                  </div>

                  <div class="text-gray-800 font-light text-sm pt-1 text-center lg:text-left px-2">
                      A-142/1, A-142, Ganesh Nagar, Tilak Nagar, New Delhi, 110018
                  </div>
              </div>
              <div class="flex flex-row items-center w-full lg:w-1/3 bg-white lg:justify-end justify-center px-2 py-4 lg:px-0 rounded-lg">
                  <span class="tracking-wider text-gray-600 cursor-pointer bg-gray-100 hover:bg-gray-200 px-4 py-2 text-sm rounded-lg leading-loose mx-2">
                      Going
                  </span>
              </div>
          </div>
      </div>
  </body>
  </html>`;
  const inputPath = path.join(folder, 'index.html');
  outputPath = path.join(folder, outputPath);
  fs.writeFileSync(inputPath, template);
  fs.writeFileSync(outputPath, '');
  return { html: inputPath, css: outputPath };
}
