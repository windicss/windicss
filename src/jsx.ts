/* eslint-disable @typescript-eslint/no-empty-interface */
export type UtilityNames =
  | 'p'
  | 'm'
  | 'w'
  | 'h'
  | 'z'
  | 'border'
  | 'grid'
  | 'flex'
  | 'bg'
  | 'text'
  | 'font'
  | 'opacity'
  | 'animate'
  | 'transition'
  | 'transform'
  | 'align'
  | 'justify'
  | 'items'
  | 'block'
  | 'content'
  | 'pos'
  | 'box'
  | 'overflow'
  | 'underline'
  | 'list'
  | 'gradient'
  | 'divide'
  | 'gap'
  | 'ring'
  | 'icon'
  | 'container'
  | 'space'
  | 'table'
  | 'order'
  | 'place'
  | 'display'
  | 'shadow'
  | 'blend'
  | 'filter'
  | 'backdrop'
  | 'cursor'
  | 'outline'
  | 'select'

export type VariantNames =
  | 'hover'
  | 'active'
  | 'focus'
  | 'enabled'
  | 'dark'
  | 'light'
  | 'sm'
  | 'lg'
  | 'md'
  | 'xl'
  | 'xxl'
  | 'first'
  | 'last'
  | 'child'
  | 'root'
  | 'before'
  | 'after'
  | 'all'

export type AttributifyNames<Prefix extends string = ''> =
  | `${Prefix}${UtilityNames}`
  | `${Prefix}${VariantNames}`
  | `${Prefix}${VariantNames}:${UtilityNames}`

export interface AttributifyAttributes extends Partial<Record<AttributifyNames, string>> {}
