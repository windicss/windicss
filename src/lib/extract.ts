import { Utility } from './utilities/handler';
import { deepCopy } from '../utils/tools';
import { Style, Property } from '../utils/style';
import { pluginOrder } from '../config/order';
import { staticUtilities, dynamicUtilities } from './utilities';
import type { Processor } from './index';

export function generateStaticStyle(processor: Processor, className:string, addComment = false): Style | undefined {
  // eslint-disable-next-line no-prototype-builtins
  if (!staticUtilities.hasOwnProperty(className))
    return;

  const style = new Style('.' + className);
  const comment = addComment ? className : undefined;
  const { utility, meta } = staticUtilities[className];
  for (const [key, value] of Object.entries(utility)) {
    style.add(
      Array.isArray(value)
        ? value.map((i) => new Property(key, i, comment))
        : new Property(key, value, comment)
    );
  }
  if (processor._plugin.core && !processor._plugin.core[meta.group]) return;
  return style.updateMeta('utilities', meta.group, pluginOrder[meta.group as keyof typeof pluginOrder], meta.order, true);
}

export default function extract(
  processor: Processor,
  className: string,
  addComment = false,
  prefix?: string,
): Style | Style[] | undefined {
  if (prefix && !className.startsWith(prefix)) return;

  // handle static base utilities
  if (!prefix && className in staticUtilities) return generateStaticStyle(processor, className, addComment);
  if (prefix && className.startsWith(prefix)) {
    className = className.replace(new RegExp(`^${prefix}`), '');
    if (className in staticUtilities) return generateStaticStyle(processor, className, addComment);
  }
  // handle static plugin utilities & components
  const staticPlugins = { ...processor._plugin.utilities, ...processor._plugin.components, ...processor._plugin.shortcuts };
  if (className in staticPlugins) return deepCopy(staticPlugins[className]);

  const utility = new Utility(className, processor._handler);


  // handle dynamic plugin utilities
  for (const [key, generator] of Object.entries(processor._plugin.dynamic)) {
    if (className.match(new RegExp(`^-?${key}`))) {
      let style = generator(utility);
      if (style instanceof Property) style = style.toStyle(utility.class);
      if (style && addComment)
        Array.isArray(style)
          ? style.map((i) => i.property.forEach((p) => (p.comment = className)))
          : style.property.forEach((p) => (p.comment = className));
      if (style) return style;
    }
  }

  // handle dynamic base utilities
  const matches = className.match(/\w+/);
  const key = matches ? matches[0] : undefined;
  // eslint-disable-next-line no-prototype-builtins
  if (key && dynamicUtilities.hasOwnProperty(key)) {
    let style = dynamicUtilities[key](utility, processor.pluginUtils);
    if (!style) return;
    if (processor._plugin.core && !processor._plugin.core[Array.isArray(style) ? style[0].meta.group : style.meta.group]) return;
    if (style instanceof Property) style = style.toStyle(utility.class);
    if (addComment) Array.isArray(style)? style.map((i) => i.property.forEach((p) => (p.comment = className))): style.property.forEach((p) => (p.comment = className));
    return style;
  }
}
