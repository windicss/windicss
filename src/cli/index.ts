import arg from 'arg';
import { deepCopy, Console } from '../utils/tools';
import { resolve, dirname, join, extname } from 'path';
import { Processor } from '../lib';
import { readFileSync, writeFile, watch, unwatchFile, existsSync } from 'fs';
import { HTMLParser, CSSParser } from '../utils/parser';
import { StyleSheet } from '../utils/style';
import {
  getVersion,
  globArray,
  generateTemplate,
  fuzzy,
} from './utils';
import type { Extractor } from '../interfaces';

const doc = `Generate css from text files that containing windi classes.
By default, it will use interpretation mode to generate a single css file.

Usage:
  windicss [filenames]
  windicss [filenames] -c -m -d
  windicss [filenames] -c -s -m -d
  windicss [filenames] [-c | -i] [-a] [-b | -s] [-m] [-d] [-p <prefix:string>] [-o <path:string>] [--args arguments]

Options:
  -h, --help            Print this help message and exit.
  -v, --version         Print windicss current version and exit.

  -i, --interpret       Interpretation mode, generate class selectors. This is the default behavior.
  -c, --compile         Compilation mode, combine the class name in each row into a single class.
  -a, --attributify     Attributify mode, generate attribute selectors. Attributify mode can be mixed with the other two modes.
  -t, --preflight       Add preflights, default is false.

  -b, --combine         Combine all css into one single file. This is the default behavior.
  -s, --separate        Generate a separate css file for each input file.

  -d, --dev             Enable hot reload and watch mode.
  -m, --minify          Generate minimized css file.
  -z, --fuzzy           Enable fuzzy match, only works in interpration mode.
  -p, --prefix PREFIX   Set the css class name prefix, only valid in compilation mode. The default prefix is 'windi-'.
  -o, --output PATH     Set output css file path.
  -f, --config PATH     Set config file path.

  --style               Parse and transform windi style block.
  --init PATH           Start a new project on the path.
`;

const args = arg({
  // Types
  '--help': Boolean,
  '--version': Boolean,
  '--compile': Boolean,
  '--interpret': Boolean,
  '--attributify': Boolean,
  '--preflight': Boolean,
  '--combine': Boolean,
  '--separate': Boolean,
  '--dev': Boolean,
  '--minify': Boolean,
  '--fuzzy': Boolean,
  '--style': Boolean,
  '--init': String,
  '--prefix': String,
  '--output': String,
  '--config': String,

  // Aliases
  '-h': '--help',
  '-v': '--version',
  '-i': '--interpret',
  '-c': '--compile',
  '-a': '--attributify',
  '-t': '--preflight',
  '-b': '--combine',
  '-s': '--separate',
  '-d': '--dev',
  '-m': '--minify',
  '-p': '--prefix',
  '-o': '--output',
  '-f': '--config',
  '-z': '--fuzzy',
});

if (args['--help'] || (args._.length === 0 && Object.keys(args).length === 1)) {
  Console.log(doc);
  process.exit();
}

if (args['--version']) {
  Console.log(getVersion());
  process.exit();
}

if (args['--init']) {
  const template = generateTemplate(args['--init'], args['--output']);
  args._.push(template.html);
  args['--preflight'] = true;
  args['--output'] = template.css;
}

const configFile = args['--config'] ? resolve(args['--config']) : undefined;
let preflights: { [key:string]: StyleSheet } = {};
let styleSheets: { [key:string]: StyleSheet } = {};
let processor = new Processor(configFile ? require(configFile) : undefined);
let safelist = processor.config('safelist');

if (configFile) Console.log('Config file:', configFile);

function compile(files: string[]) {
  // compilation mode
  const prefix = args['--prefix'] ?? 'windi-';
  files.forEach((file) => {
    let indexStart = 0;
    const outputStyle: StyleSheet[] = [];
    const outputHTML: string[] = [];
    const html = readFileSync(file).toString();
    const parser = new HTMLParser(html);

    // Match ClassName then replace with new ClassName
    parser.parseClasses().forEach((p) => {
      outputHTML.push(html.substring(indexStart, p.start));
      const utility = processor.compile(p.result, prefix, true, args['--dev']); // Set third argument to false to hide comments;
      outputStyle.push(utility.styleSheet);
      outputHTML.push([utility.className, ...utility.ignored].join(' '));
      indexStart = p.end;
    });
    outputHTML.push(html.substring(indexStart));
    const added = (
      outputStyle.reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
    );
    styleSheets[file] = args['--dev'] ? (styleSheets[file]? styleSheets[file].extend(added) : added) : added;

    const outputFile = file.replace(/(?=\.\w+$)/, '.windi');
    writeFile(outputFile, outputHTML.join(''), () => null);
    Console.log(`${file} -> ${outputFile}`);
    if (args['--preflight']) {
      if (args['--dev']) {
        const preflight = processor.preflight(html, true, true, true, true);
        preflights[file] = preflights[file] ? preflights[file].extend(preflight) : preflight;
      } else {
        preflights[file] = processor.preflight(html);
      }
    }
  });
}

function interpret(files: string[]) {
  // interpretation mode
  files.forEach((file) => {
    const content = readFileSync(file).toString();
    let classes: string[] = [];
    if (args['--fuzzy']) {
      classes = fuzzy(content);
    } else {
      const parser = new HTMLParser(content);
      classes = parser.parseClasses().map((i) => i.result);
    }
    const extractors = processor.config('extract.extractors') as Extractor[] | undefined;
    if (extractors) {
      for (const { extractor, extensions } of extractors) {
        if (extensions.includes(extname(file).slice(1))) {
          const result = extractor(content);
          if ('classes' in result && result.classes) {
            classes = [...classes, ...result.classes];
          }
        }
      }
    }
    if (args['--dev']) {
      const utility = processor.interpret(classes.join(' '), true);
      styleSheets[file] = styleSheets[file] ? styleSheets[file].extend(utility.styleSheet) : utility.styleSheet;
      if (args['--preflight']) {
        const preflight = processor.preflight(content, true, true, true, true);
        preflights[file] = preflights[file] ? preflights[file].extend(preflight) : preflight;
      }
    } else {
      const utility = processor.interpret(classes.join(' '));
      styleSheets[file] = utility.styleSheet;
      if (args['--preflight']) preflights[file] = processor.preflight(content);
    }
  });
}

function attributify(files: string[]) {
  // attributify mode
  files.forEach((file) => {
    const parser = new HTMLParser(readFileSync(file).toString());
    const attrs: { [key: string]: string | string[] } = parser
      .parseAttrs()
      .reduceRight((a: { [key: string]: string | string[] }, b) => {
        if (b.key === 'class' || b.key === 'className') return a;
        if (b.key in a) {
          a[b.key] = Array.isArray(a[b.key])
            ? Array.isArray(b.value)? [ ...(a[b.key] as string[]), ...b.value ]: [ ...(a[b.key] as string[]), b.value ]
            : [ a[b.key] as string, ...(Array.isArray(b.value) ? b.value : [b.value]) ];
          return a;
        }
        return Object.assign(a, { [b.key]: b.value });
      }, {});
    if (args['--dev']) {
      const utility = processor.attributify(attrs, true);
      styleSheets[file] = styleSheets[file] ? styleSheets[file].extend(utility.styleSheet) : utility.styleSheet;
    } else {
      const utility = processor.attributify(attrs);
      styleSheets[file] = utility.styleSheet;
    }
  });
}

function styleBlock(files: string[]) {
  files.forEach((file) => {
    const content = readFileSync(file).toString();
    const block = content.match(/(?<=<style lang=['"]windi["']>)[\s\S]*(?=<\/style>)/);
    if (block && block.index) {
      const css = content.slice(block.index, block.index + block[0].length);
      const parser = new CSSParser(css, processor);
      styleSheets[file] = styleSheets[file].extend(parser.parse());
    }
  });
}

function build(files: string[], update = false) {
  if (args['--compile']) {
    compile(files);
  } else {
    interpret(files);
  }
  if (args['--attributify']) attributify(files);
  if (args['--style']) styleBlock(files);
  if (args['--separate']) {
    for (const [file, sheet] of Object.entries(styleSheets)) {
      const outfile = file.replace(/\.\w+$/, '.windi.css');
      writeFile(outfile, (args['--preflight'] ? deepCopy(sheet).extend(preflights[file], false) : sheet).build(args['--minify']), () => null);
      Console.log(`${file} -> ${outfile}`);
    }
  } else {
    let outputStyle = Object.values(styleSheets)
      .reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
      .sort()
      .combine();
    if (args['--preflight'])
      outputStyle = Object.values(preflights)
        .reduce(
          (previousValue: StyleSheet, currentValue: StyleSheet) =>
            previousValue.extend(currentValue),
          new StyleSheet()
        )
        .sort()
        .combine()
        .extend(outputStyle);
    const filePath = args['--output'] ?? 'windi.css';
    writeFile(filePath, outputStyle.build(args['--minify']), () => null);
    if (!update) {
      Console.log('Matched files:', files);
      Console.log('Output file:', resolve(filePath));
    }
  }

}

function buildSafeList(safelist: unknown) {
  if (safelist) {
    let classes: string[] = [];
    if (typeof safelist === 'string') {
      classes = safelist.split(/\s/).filter(i => i);
    }
    if (Array.isArray(safelist)) {
      for (const item of safelist) {
        if (typeof item === 'string') {
          classes.push(item);
        } else if (Array.isArray(item)) {
          classes = classes.concat(item);
        }
      }
    }
    styleSheets['safelist'] = processor.interpret(classes.join(' ')).styleSheet;
  }
}

const patterns = args._
  .concat(processor.config('extract.include', []) as string[])
  .concat((processor.config('extract.exclude', []) as string[]).map(i => '!' + i));

let matchFiles = globArray(patterns);

if (matchFiles.length === 0) {
  Console.error('No files were matched!');
  process.exit();
}

buildSafeList(safelist);
build(matchFiles);

function watchBuild(file: string) {
  watch(file, (event, path) => {
    if (event === 'rename') {
      const newFiles = globArray(patterns);
      const renamed = matchFiles.filter(i => !(newFiles.includes(i)))[0];
      if (existsSync(path)) {
        Console.log('File', `'${renamed}'`, 'has been renamed to', `'${path}'`);
        matchFiles = newFiles;
        Console.log('Matched files:', matchFiles);
      } else {
        Console.log('File', `'${file}'`, 'has been deleted');
        unwatchFile(file);
        matchFiles = newFiles;
        delete styleSheets[file];
        delete preflights[file];
        if (matchFiles.length > 0) {
          Console.log('Matched files:', matchFiles);
          Console.time('Building');
        }
        build([], true);
        if (matchFiles.length > 0) {
          Console.timeEnd('Building');
        } else {
          Console.error('No files were matched!');
          process.exit();
        }
      }
    }
    if (event === 'change') {
      Console.log('File', `'${path}'`, 'has been changed');
      Console.time('Building');
      build([file], true);
      Console.timeEnd('Building');
    }
  });
}

function watchConfig(file?: string) {
  if (!file) return;
  let stamp = 0;
  watch(file, (event, path) => {
    if (event === 'change' && (stamp === 0 || + new Date() - stamp > 500)) {
      // fix fire twice event when change config file
      stamp = + new Date();
      Console.log('Config', `'${path}'`, 'has been changed');
      Console.time('Building');
      configFile && delete require.cache[configFile];
      processor = new Processor(configFile ? require(configFile) : undefined);
      safelist = processor.config('safelist');
      styleSheets = {};
      preflights = {};
      buildSafeList(safelist);
      build(matchFiles, true);
      Console.timeEnd('Building');
    }
  });
}

if (args['--dev']) {
  watchConfig(configFile);
  for (const file of matchFiles) {
    watchBuild(file);
  }
  for (const dir of Array.from(new Set(matchFiles.map(f => dirname(f))))) {
    watch(dir, (event, path) => {
      if (event === 'rename' && existsSync(join(dir, path))) {
        // when create new file
        const newFiles = globArray(patterns);
        if (newFiles.length > matchFiles.length) {
          const newFile = newFiles.filter(i => !matchFiles.includes(i))[0];
          Console.log('New file', `'${newFile}'`,  'added');
          matchFiles.push(newFile);
          Console.log('Matched files:', matchFiles);
          Console.time('Building');
          build([newFile], true);
          watchBuild(newFile);
          Console.timeEnd('Building');
        }
      }
    });
  }
}
