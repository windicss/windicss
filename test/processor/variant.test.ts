import { Property } from '../../src/utils/style';
import type { Style } from '../../src/utils/style';
import { baseConfig } from '../../src/config';
import {
  generateOrientations,
  generateScreens,
  generateStates,
  generateThemes,
  resolveVariants,
} from '../../src/lib/variants';
import { Processor } from '../../src/lib';

function _generateTestVariants(variants: { [key: string]: () => Style }) {
  const output: { [key: string]: string } = {};
  for (const [name, func] of Object.entries(variants)) {
    const style = func();
    style.selector = '.test';
    style.add(new Property('background', '#1C1C1E'));
    output[name] = style.build();
  }
  return output;
}

describe('Variants', () => {
  it('generate screens', () => {
    const screens = generateScreens({
      sm: '640px',
      lg: '1024px',
      print: { raw: 'print' },
      narrow: { max: '768px' },
    });
    expect(_generateTestVariants(screens)).toMatchSnapshot('vars');

    const unsortedScreens = generateScreens({
      print: { raw: 'print' },
      lg: '1024px',
      sm: '640px',
      narrow: { max: '768px' },
    });
    expect(Object.keys(unsortedScreens)).toMatchSnapshot('screens');
  });

  it('generate themes with darkMode class', () => {
    const themes = generateThemes('class');
    expect(_generateTestVariants(themes)).toMatchSnapshot('vars');
  });

  it('generate themes with darkMode media', () => {
    const themes = generateThemes('media');
    expect(_generateTestVariants(themes)).toMatchSnapshot('vars');
  });

  it('generate states', () => {
    const allStates = generateStates(baseConfig.variantOrder ?? []);
    expect(Object.keys(allStates)).toEqual(baseConfig.variantOrder ?? []);

    expect(_generateTestVariants(allStates)).toMatchSnapshot('all');

    const someStates = generateStates([
      'focus',
      'hover',
      'before',
      'svg',
      'group-hover',
      'motion-safe',
      'wrong',
    ]);
    expect(_generateTestVariants(someStates)).toMatchSnapshot('some');
  });

  it('resolve variants', () => {
    const variants = resolveVariants(baseConfig);
    expect(Object.keys(variants)).toEqual(['orientation', 'screen', 'theme', 'state']);
    expect(Object.keys(variants.screen)).toMatchSnapshot('screen');
    expect(Object.keys(variants.theme)).toEqual([
      '@dark',
      '@light',
      '.dark',
      '.light',
      'dark',
      'light',
    ]);
    expect(Object.keys(variants.state)).toEqual(baseConfig.variantOrder ?? []);
    expect(Object.keys(variants.orientation)).toEqual(['portrait', 'landscape']);

    const emptyVariants = resolveVariants({});
    expect(emptyVariants.orientation).toEqual({});
    expect(emptyVariants.screen).toEqual({});
    expect(emptyVariants.theme).toEqual({});
    expect(emptyVariants.state).toEqual({});
  });

  it('should not generate utility when variant is wrong', () => {
    const processor = new Processor();
    expect(processor.interpret('pointer-hover:bg-gray-200 hover:cursor:bg-gray-200').ignored.length).toEqual(2);
  });

  it('directions', () => {
    const processor = new Processor();
    expect(processor.interpret('ltr:text-lg rtl:dark:text-sm').styleSheet.build()).toMatchSnapshot('css');
  });

  it('orientation', () => {
    const orientations = generateOrientations({
      portrait: 'portrait',
      landscape: 'landscape',
    });
    expect(_generateTestVariants(orientations)).toMatchSnapshot('css');
  });
});
