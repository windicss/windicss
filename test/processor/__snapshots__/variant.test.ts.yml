Tools / directions / css / 0: |-
  [dir='ltr'] .ltr\:text-lg, [dir='ltr'].ltr\:text-lg {
    font-size: 1.125rem;
    line-height: 1.75rem;
  }
  [dir='rtl'] .dark .rtl\:dark\:text-sm, [dir='rtl'].dark .rtl\:dark\:text-sm {
    font-size: 0.875rem;
    line-height: 1.25rem;
  }
Tools / generate screens / screens / 1: |-
  [
    "print",
    "narrow",
    "sm",
    "<sm",
    "@sm",
    "-sm",
    "+sm",
    "lg",
    "<lg",
    "@lg",
    "-lg",
    "+lg"
  ]
Tools / generate screens / vars / 0: |-
  {
    "print": "@media print {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "narrow": "@media (max-width: 768px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "sm": "@media (min-width: 640px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "<sm": "@media (max-width: 639.9px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "@sm": "@media (min-width: 640px) and (max-width: 1023.9px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "-sm": "@media (max-width: 640px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "+sm": "@media (min-width: 640px) and (max-width: 1024px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "lg": "@media (min-width: 1024px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "<lg": "@media (max-width: 1023.9px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "@lg": "@media (min-width: 1024px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "-lg": "@media (max-width: 1024px) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "+lg": "@media (min-width: 1024px) {\n  .test {\n    background: #1C1C1E;\n  }\n}"
  }
Tools / generate states / all / 0: |-
  {
    "hover": ".test:hover {\n  background: #1C1C1E;\n}",
    "focus": ".test:focus {\n  background: #1C1C1E;\n}",
    "active": ".test:active {\n  background: #1C1C1E;\n}",
    "visited": ".test:visited {\n  background: #1C1C1E;\n}",
    "link": ".test:link {\n  background: #1C1C1E;\n}",
    "target": ".test:target {\n  background: #1C1C1E;\n}",
    "focus-visible": ".test:focus-visible {\n  background: #1C1C1E;\n}",
    "focus-within": ".test:focus-within {\n  background: #1C1C1E;\n}",
    "checked": ".test:checked {\n  background: #1C1C1E;\n}",
    "not-checked": ".test:not(:checked) {\n  background: #1C1C1E;\n}",
    "default": ".test:default {\n  background: #1C1C1E;\n}",
    "disabled": ".test:disabled {\n  background: #1C1C1E;\n}",
    "enabled": ".test:enabled {\n  background: #1C1C1E;\n}",
    "indeterminate": ".test:indeterminate {\n  background: #1C1C1E;\n}",
    "invalid": ".test:invalid {\n  background: #1C1C1E;\n}",
    "valid": ".test:valid {\n  background: #1C1C1E;\n}",
    "optional": ".test:optional {\n  background: #1C1C1E;\n}",
    "required": ".test:required {\n  background: #1C1C1E;\n}",
    "placeholder-shown": ".test:placeholder-shown {\n  background: #1C1C1E;\n}",
    "read-only": ".test:read-only {\n  background: #1C1C1E;\n}",
    "read-write": ".test:read-write {\n  background: #1C1C1E;\n}",
    "not-disabled": ".test:not(:disabled) {\n  background: #1C1C1E;\n}",
    "first-of-type": ".test:first-of-type {\n  background: #1C1C1E;\n}",
    "not-first-of-type": ".test:not(:first-of-type) {\n  background: #1C1C1E;\n}",
    "last-of-type": ".test:last-of-type {\n  background: #1C1C1E;\n}",
    "not-last-of-type": ".test:not(:last-of-type) {\n  background: #1C1C1E;\n}",
    "first": ".test:first-child {\n  background: #1C1C1E;\n}",
    "last": ".test:last-child {\n  background: #1C1C1E;\n}",
    "not-first": ".test:not(:first-child) {\n  background: #1C1C1E;\n}",
    "not-last": ".test:not(:last-child) {\n  background: #1C1C1E;\n}",
    "only-child": ".test:only-child {\n  background: #1C1C1E;\n}",
    "not-only-child": ".test:not(:only-child) {\n  background: #1C1C1E;\n}",
    "only-of-type": ".test:only-of-type {\n  background: #1C1C1E;\n}",
    "not-only-of-type": ".test:not(:only-of-type) {\n  background: #1C1C1E;\n}",
    "even": ".test:nth-child(even) {\n  background: #1C1C1E;\n}",
    "odd": ".test:nth-child(odd) {\n  background: #1C1C1E;\n}",
    "even-of-type": ".test:nth-of-type(even) {\n  background: #1C1C1E;\n}",
    "odd-of-type": ".test:nth-of-type(odd) {\n  background: #1C1C1E;\n}",
    "root": ".test:root {\n  background: #1C1C1E;\n}",
    "empty": ".test:empty {\n  background: #1C1C1E;\n}",
    "before": ".test::before {\n  background: #1C1C1E;\n}",
    "after": ".test::after {\n  background: #1C1C1E;\n}",
    "first-letter": ".test::first-letter {\n  background: #1C1C1E;\n}",
    "first-line": ".test::first-line {\n  background: #1C1C1E;\n}",
    "file-selector-button": ".test::file-selector-button {\n  background: #1C1C1E;\n}",
    "file": ".test::file-selector-button {\n  background: #1C1C1E;\n}",
    "selection": ".test::selection {\n  background: #1C1C1E;\n}",
    "marker": ".test *::marker, .test::marker {\n  background: #1C1C1E;\n}",
    "svg": ".test svg {\n  background: #1C1C1E;\n}",
    "all": ".test * {\n  background: #1C1C1E;\n}",
    "children": ".test > * {\n  background: #1C1C1E;\n}",
    "siblings": ".test ~ * {\n  background: #1C1C1E;\n}",
    "sibling": ".test + * {\n  background: #1C1C1E;\n}",
    "ltr": "[dir='ltr'] .test, [dir='ltr'].test {\n  background: #1C1C1E;\n}",
    "rtl": "[dir='rtl'] .test, [dir='rtl'].test {\n  background: #1C1C1E;\n}",
    "group-hover": ".group:hover .test {\n  background: #1C1C1E;\n}",
    "group-focus": ".group:focus .test {\n  background: #1C1C1E;\n}",
    "group-active": ".group:active .test {\n  background: #1C1C1E;\n}",
    "group-visited": ".group:visited .test {\n  background: #1C1C1E;\n}",
    "motion-safe": "@media (prefers-reduced-motion: no-preference) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "motion-reduce": "@media (prefers-reduced-motion: reduce) {\n  .test {\n    background: #1C1C1E;\n  }\n}"
  }
Tools / generate states / some / 1: |-
  {
    "focus": ".test:focus {\n  background: #1C1C1E;\n}",
    "hover": ".test:hover {\n  background: #1C1C1E;\n}",
    "before": ".test::before {\n  background: #1C1C1E;\n}",
    "svg": ".test svg {\n  background: #1C1C1E;\n}",
    "group-hover": ".group:hover .test {\n  background: #1C1C1E;\n}",
    "motion-safe": "@media (prefers-reduced-motion: no-preference) {\n  .test {\n    background: #1C1C1E;\n  }\n}"
  }
Tools / generate themes with darkMode class / vars / 0: |-
  {
    "@dark": "@media (prefers-color-scheme: dark) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "@light": "@media (prefers-color-scheme: light) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    ".dark": ".dark .test {\n  background: #1C1C1E;\n}",
    ".light": ".light .test {\n  background: #1C1C1E;\n}",
    "dark": ".dark .test {\n  background: #1C1C1E;\n}",
    "light": ".light .test {\n  background: #1C1C1E;\n}"
  }
Tools / generate themes with darkMode media / vars / 0: |-
  {
    "@dark": "@media (prefers-color-scheme: dark) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "@light": "@media (prefers-color-scheme: light) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    ".dark": ".dark .test {\n  background: #1C1C1E;\n}",
    ".light": ".light .test {\n  background: #1C1C1E;\n}",
    "dark": "@media (prefers-color-scheme: dark) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "light": "@media (prefers-color-scheme: light) {\n  .test {\n    background: #1C1C1E;\n  }\n}"
  }
Tools / orientation / css / 0: |-
  {
    "portrait": "@media (orientation: portrait) {\n  .test {\n    background: #1C1C1E;\n  }\n}",
    "landscape": "@media (orientation: landscape) {\n  .test {\n    background: #1C1C1E;\n  }\n}"
  }
Tools / resolve variants / screen / 0: |-
  [
    "sm",
    "<sm",
    "@sm",
    "-sm",
    "+sm",
    "md",
    "<md",
    "@md",
    "-md",
    "+md",
    "lg",
    "<lg",
    "@lg",
    "-lg",
    "+lg",
    "xl",
    "<xl",
    "@xl",
    "-xl",
    "+xl",
    "2xl",
    "<2xl",
    "@2xl",
    "-2xl",
    "+2xl"
  ]
